WIP: Provider::Package to use EVRAs for dnf rpm version comparisons

epilatow opened the pull request:
This is a potential bug fix for [Issue 9940](https://github.com/chef/chef/issues/9940).

Currently on dnf systems, Provider::Package does rpm version comparisons
between installed_version/candidate_version (EVRAs) and
new_resource.version / new_version_array (a VR or EVR). These
comparisons can fail in dnf_helper.py because versioncompare doesn't
know the type of version data it's comparing (and it's impossible to
deterministically parse an EVR[A] string without first knowing if it's
an EVR or EVRA).

To fix this problem, we first override new_version_array in
Chef::Provider::Package to just return nil values. This forces the
Package class to use candidate_version in place of new_version_array for
all version comparisons.

This doesn't fix rpm version comparisons in and of itself, but instead
it ensures that all version comparisons are done using EVRAs. This
allows us to update dnf_python.py versioncompare to properly parse and
compare EVRAs. This fixes rpm version comparisons for package resources
which contain a version specification and an :upgrade action.

To fix rpm version comparisons for package resources that have a version
specification and an :install action, we also need to update
target_version_array in Chef::Provider::Package to fallback to comparing
against candidate_version when new_version_array returns nil.

To test this fix I created four packages, each with two versions:
2.02-78.el8 and 2.02-78.el8_1.1.  I installed the 2.02-78.el8 version of
all the packages, and then executed a chef run with the following
resources configured:

---8<---
package 'empty-test-rpm-version-upgrade' do
  version '1:2.02-78.el8_1.1'
  action :upgrade
end

package 'empty-test-rpm-version-install' do
  version '1:2.02-78.el8_1.1'
  action :install
end

package 'empty-test-rpm-name-upgrade-1:2.02-78.el8_1.1' do
  action :upgrade
end

package 'empty-test-rpm-name-install-1:2.02-78.el8_1.1' do
  action :install
end
---8<---

and then I confirmed that all the packages were upgraded as expected.

I also regression tested that if an invalid new_version is requested
(ie, a version for which we can't find a candidate_version) that chef
fails (as expected).

Signed-off-by: Edward Pilatowicz <epilatow@fb.com>

jaymzh commented on lib/chef/provider/package.rb:
```diff
@@ -452,9 +452,15 @@ def target_version_array
+                elsif !allow_downgrade && version_compare(current_version, candidate_version) == 1
+                  logger.trace("#{new_resource} #{package_name} has installed version #{current_version}, which is newer than available version #{candidate_version}. Skipping...)")
+                  target_version_array.push(nil)
+                else
```
`:install` doesn't upgrade like this. Unless the version is specified, `:install` simply installs if nothing is installed. You need to use `:upgrade` to upgrade-to-latest or `:install` with a version to move to a new version.

epilatow replied:
Thanks  for looking at this.  Also, thanks for the description of the required behavior (I'm new to chef and didn't know this).

I added this code to fix the following test case:
---8<---
package 'empty-test-rpm-version-install' do
version '1:2.02-78.el8_1.1'
action :install
end
---8<---

With this change, new_version (which contained the user specified version) has been nil'd out (by the dnf subclass).  So I no longer have the context necessary to make the determination if we should do the upgrade in the scenario you've outlined.  :(

I'm hesitant to update package.rb to add special handling for this case since it's a generic resource handler and this information is only needed for this one dnf edge case.

My first idea for a fix would probably be to update the dnf version of new_version_array() such that: instead of always nil'ing out the entire array, if we're dealing with an install package resource, then we could set new_version to the available_version if both are non-nil (and set it to nil otherwise).

I'd be curious to get @lamont-granquist perspective on this idea (and the rest of the diff).  I noticed that he just closed the issue this diff is addressing as a dup of another issue+PR, so I suspect this PR may be obsoleted shortly.

epilatow replied:
I've updated the PR with the changes I described above, which allowed me to revert all the updates to provider/package.rb (so I've restored the old install behavior).  I've verified that all my original tests (from the comments) pass.

epilatow commented:
We've been running with these dnf changes for since June, but last week we discovered a bug in them. Specifically, by nil'ing out new_versions when there is no candidate_version, chef would succeed if a package resource specified an invalid version. (Ie, when a requested version couldn't be mapped to a candidate version.) I've just updated this PR to address the issue. This required leaving new_versions alone if there was no candidate version, which complicated things because we can't use unresolved new_versions strings for comparisons against candidate versions (ie full EVRAs). So I added tagging for EVRAs so we can ensure that we don't accidentally compare them to non-EVRAs.

lamont-granquist commented:
> Specifically, by nil'ing out new_versions when there is no candidate_version, chef would succeed if a package resource specified an invalid version. 

yeah, i think that's why this PR stalled so hard because I knew that had to be wrong...

epilatow commented:
Obsoleted by PR #9847
